### DISCUSSION TITLE
How to configure BLE Application with multiple services

### DISCUSSION THREAD

**Discussion by bem22** [zephyrproject-rtos/zephyr#61285] [Q&A] [answered]
I'm writing here because I exhausted other resources (nordic forums, chat gpt, google and datasheets)

I'm trying to build a Bluetooth LE application for the nRF52840 chip. So far it was easy to understand from examples how to configure the GATT profile while using a single custom service (with custom UUID) - but now I would like to structure my application such that the service is not cluttered with characteristics that don't belong in the same category (for example I want a service that provides raw data from sensors and one that provides computed sensor outputs), a service for button controls and a service for controlling an LED array.

I can't wrap my head around the concept of how services are discovered and whether every service has to be shown **at advertising time** or whether there is a possibility to discover the services AFTER the connection is done with the device - to my understanding this is what happens with the rest of attributes. But this could be wrong. Please correct me.

Now, IF the first option is right, and all services have to be shown at advertising time, then how does **extended advertising work** and how can uplift the limit of 31 bytes (at most one 128 bit/16 bytes vendor specific UUIDs will fit in a 31 byte packet)? 

In my search for answering this question I stumbled upon [this broadcaster_multiple example](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/bluetooth/broadcaster_multiple) and **[this multiple advertisement sets example](https://github.com/nrfconnect/sdk-nrf/blob/main/samples/bluetooth/multiple_adv_sets/src/main.c)**, but neither of them quite does what I would like it to do, which is channeling multiple services with their respective UUIDs into one device.

Here is an example from a sone Bluetooth device I own and is the functionality I want to replicate:
<img src="https://github.com/zephyrproject-rtos/zephyr/assets/29655042/e3b7e479-6c6d-488e-a44a-192344921c6f" alt="drawing" width="350"/>)

Thank you for reading until the end.

**Comment by hermabe** [comment]
The GATT server can have any number of services, each containing any number of characteristics. You can use multiple `BT_GATT_SERVICE_DEFINE` or multiple calls to `bt_gatt_service_register` to add multiple services.

The services present in the GATT server are not required to be present in the advertising data. Some standardized profiles may require this but you can do as you want if you are defining your own service. If you add the UUIDs to the advertising data you can avoid connecting to devices that do not have the services you are looking for.

In any case you will have to do service discovery once connected to get the handles of the attributes. This is normally only required the first time you connect to the server, the layout of the GATT database usually doesn't change. You can subscribe to "Service Changed Indications" or read the "Database Hash Characteristic" to know if the database has changed and re-discovery is required.

**Reply by bem22** [reply]
So I used BT_GATT_SERVICE_DEFINE twice in my app but it will not present two services.

I have the following code:
```
/* LED Button Service Declaration */
BT_GATT_SERVICE_DEFINE(my_lbs_svc,
BT_GATT_PRIMARY_SERVICE(BT_UUID_LBS),
/* STEP 1 - Modify the Button characteristic declaration to support indication */
	BT_GATT_CHARACTERISTIC(BT_UUID_LBS_BUTTON,
			       BT_GATT_CHRC_READ | BT_GATT_CHRC_INDICATE,
			       BT_GATT_PERM_READ, read_button, NULL,
			       &button_state),
/* STEP 2 - Create and add the Client Characteristic Configuration Descriptor */
	BT_GATT_CCC(my_lbs_cb_ccc_cfg_changed,
				   BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),	

	BT_GATT_CHARACTERISTIC(BT_UUID_LBS_LED,
			       BT_GATT_CHRC_WRITE,
			       BT_GATT_PERM_WRITE,
			       NULL, write_led, NULL),

);
```

This service shows up as intended and is also present in the advertisement packet (if I understand this correctly?) through its reference:
```
static const struct bt_data sd[] = {
	BT_DATA_BYTES(BT_DATA_UUID128_SOME, BT_UUID_LBS_VAL)
};
```

This is my second service:
```
BT_GATT_SERVICE_DEFINE(my_myservice_svc,
    BT_GATT_PRIMARY_SERVICE(BT_UUID_MYSERVICE),

    	BT_GATT_CHARACTERISTIC(BT_UUID_MYSERVICE_MYSENSOR,
			       BT_GATT_CHRC_NOTIFY,
			       BT_GATT_PERM_NONE, NULL, NULL,
			       NULL),

        BT_GATT_CCC(myservice_ccc_mysensor_cfg_changed,
		    BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
);
```
This is not referenced in the advertisement, and it doesn't show.

No errors are thrown in the console either.

**Reply by hermabe** [reply, marked as answer]
Does the [peripheral sample](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/bluetooth/peripheral) work for you? Can you see all the services there (GATT, GAP, BAS, CTS, DIS, HRS, IAS, Custom/Unknown)? If you copy-paste the custom service can you see two "Unknown service"?

**Reply by bem22** [reply]
Yes, @hermabe 

Here's proof. It does what I want. I am too tired to figure out what does not work in the original app but I will debug it tomorrow and now I have a path to do it.

<img src="https://github.com/zephyrproject-rtos/zephyr/assets/29655042/56691522-b5f2-4703-ba13-f01a868544c8" alt="drawing" width="350"/>

Thank you so much for reading, understanding and then replying to my query. It means the world to me.

I have no way to thank you but with words here. Wait, there is something: If you have a youtube channel, a non-political organization or any kind of cause that you would like to donate 5 euro/usd, please name it and I will do it in your name to keep the ball rolling.

Thank you so much, again!

**Reply by bem22** [reply]
Before this post, I did not understand the fundamental concepts of how GATT profiles are built and I confused the advertisement for a requirement to add a new service. However, the reason why I didn't work is laughable.

I'm embarrassed to say that I have not given my compiler a chance to work on those two files before I added these lines highlighted:

<img src="https://github.com/zephyrproject-rtos/zephyr/assets/29655042/78e5085d-dc34-4738-96eb-8bb0513c413a" alt="drawing" width="350"/>

<img src="https://github.com/zephyrproject-rtos/zephyr/assets/29655042/5ec98711-33b3-401c-b245-d54e8f848524" alt="drawing" width="350"/>
